JITServer AOT cache deserializer

AlexeyKhrabrov opened the pull request:
This PR adds the `JITServerAOTDeserializer` class that will be used at the JIT client to deserialize cached AOT methods received from the JITServer. The deserializer updates each SCC offset in relocation and validation records stored in method relocation data such that it points to the correct entity in the local SCC, which is looked up and validated using the information in AOT serialization records attached to the serialized AOT method.

AlexeyKhrabrov commented:
@mpirvu This PR is the next item in #12153. Can I please get a review? @dsouzai FYI.

AlexeyKhrabrov commented:
There seems to be an issue in the `TR_J9SharedCache::rememberClass()` implementation that manifests in a rare race condition that leads to a fatal assertion in the AOT cache deserializer. @dsouzai Could you please give your opinion on what would be the best way to address this?

The assertion is `Shared cache pointer ... out of bounds` and it gets triggered by this call: https://github.com/eclipse/openj9/blob/23f391972c243904dcbde4c2d6df5e1fd879da17/runtime/compiler/runtime/JITServerAOTDeserializer.cpp#L519

`chain` here is the result of a successful `rememberClass()` call a few lines above, so by definition it must be in the SCC.

Debugging lead me to the conclusion that this is caused by the following race condition:
- Thread 1 calls `rememberClass()` and stores a new class chain. Since it calls `storeSharedData()` with `J9SHRDATA_SINGLE_STORE_FOR_KEY_TYPE`, it acquires the SCC write mutex in the mode that allows concurrent readers (see https://github.com/eclipse/openj9/blob/master/runtime/shared_common/CacheMap.cpp#L3891-L3900).
- Thread 2 calls `rememberClass()` for the same class and finds the class chain stored by Thread 1, concurrently with Thread 1 since it's allowed to acquire the SCC read mutex as explained above.
- After Thread 1 stores the class chain, but before it updates the `updateSRP` in the SCC header through the `commitUpdate()` call here: https://github.com/eclipse/openj9/blob/master/runtime/shared_common/CacheMap.cpp#L3822, Thread 2 tries to convert the class chain pointer to the SCC offset and hits the assertion because the `updateSRP` has not yet been updated.

Another place where this race condition could be triggered is `J9::AheadOfTimeCompile::getClassChainOffset()`: https://github.com/eclipse/openj9/blob/master/runtime/compiler/codegen/J9AheadOfTimeCompile.cpp#L47-L50.

Storing class chains with `J9SHRDATA_SINGLE_STORE_FOR_KEY_TYPE_OVERWRITE` would fix the race condition, but I'm not absolutely sure if it could somehow break correctness of AOT. The key that a class chain is stored under is the SCC offset of the ROMClass. This makes me think that being able to overwrite the value for this key would actually reduce AOT load failures in cases when the ROMClass for one of the superclasses changes but the "leaf" ROMClass stays the same. Calling `rememberClass()` for the new version will return the existing class chain which doesn't match the class anymore, so any AOT methods that depend on it will never be loaded. Please correct me if my understanding is wrong.

AlexeyKhrabrov commented:
The race condition described above can also potentially happen when storing a "well-known classes" object in the SCC during deserialization (although it's a lot less likely):
https://github.com/eclipse/openj9/blob/23f391972c243904dcbde4c2d6df5e1fd879da17/runtime/compiler/runtime/JITServerAOTDeserializer.cpp#L595-L603

Note that the race condition doesn't apply to existing code because the WKC object is stored in `populateWellKnownClasses()` at the beginning of the compilation, and its pointer is converted to the SCC offset much later in `J9::<arch>::AheadOfTimeCompile::processRelocations()`.

As far as I understand, `flags=J9SHRDATA_SINGLE_STORE_FOR_KEY_TYPE_OVERWRITE` and `flags=0` (which is used for well-known classes in existing code) are essentially equivalent. The only differences are overwriting in place vs. marking existing version as stale (which look the same for the readers), and the locking behaviour. I think we can safely switch to `J9SHRDATA_SINGLE_STORE_FOR_KEY_TYPE_OVERWRITE` here.

dsouzai commented:
I don't think using `J9SHRDATA_SINGLE_STORE_FOR_KEY_TYPE_OVERWRITE` is the way to go because it means we will always write to the SCC every time we call `rememberClass`. One thing I'm unsure about is 

>Thread 2 calls rememberClass() for the same class and finds the class chain stored by Thread 1, concurrently with Thread 1 since it's allowed to acquire the SCC read mutex as explained above.

If the header hasn't been updated yet, how does the SCC API find the chain? I'm not asking because I think what you're doing is wrong, I'm just surprised that the SCC API is able to find the chain even though the bounds of the SCC hasn't been updated yet.

@hangshao0 Why would `findSharedData` with `J9SHR_DATA_TYPE_AOTCLASSCHAIN` return a value if `commitUpdate` hasn't finished yet? If that's possible, shouldn't we not do that? Otherwise we're going to run into race conditions as seen in this PR.

hangshao0 commented:
The sequence of the operations is: 
1. Store the data to the SCC by memcpy()
2. Add it to the hashtable by localBDM->storeNew().
3. Update the updateSRP in the cache header by commitUpdate(). (Header is updated after 1 and 2 because we want to make sure store has been successfully done). 
https://github.com/eclipse/openj9/blob/834cc45a3e7abc4ff571e5c4f4a2258b43110093/runtime/shared_common/CacheMap.cpp#L3809-L3822

Once it is in the hashtable, it can be found. The data is already successfully stored in the SCC at that point. 

> Thread 2 tries to convert the class chain pointer to the SCC offset and hits the assertion because the updateSRP has not yet been updated.

updateSRP is not guaranteed to be updated because of the timing hole between 2 and 3.

hangshao0 commented:
The assertion on the updateSRP is not correct based on the current implementation. But it looks like we can switch 2 and 3, if you want to keep that assertion.

dsouzai commented:
> But it looks like we can switch 2 and 3, if you want to keep that assertion.

I don't know if that's a good idea, I think the fact that 3 only gets updated after a store is successful is the right thing to do. It's unfortunate on the JIT side because using the `offsetInSharedCacheFromPointer` API is a very efficient means to get the offset from a pointer in (the metadata section of) the SCC.

I'll have to think some more about what the JIT should do. We might just need to synchronize on reads/writes from/to the SCC.

dsouzai commented:
The only other option that would make the `storeSharedData` API stronger would be if it held the SCC write lock until `commitUpdate` finished. Is that doable @hangshao0 ?

AlexeyKhrabrov commented:
There are 2 easy workaround options that the JIT can do to avoid the fatal assertion:
- Fail the AOT compilation or load instead of the assertion. This race condition is relatively rare, so performance impact should be small.
- Retry the pointer-offset conversion a few times until it succeeds (which is guaranteed to happen). This is ugly, but seems to work well.

dsouzai commented:
I don't prefer either of them but especially not the second option. Technically this is possible even without JITServer; we just at the moment don't try to get the offset immediately after rememberClass. Ideally the best thing to do would be hold the write lock until the update was complete to prevent readers, but we'll see what Hang says about that. In terms of JITServer, you could synchronize around the call to
```
uintptr_t *chain = _sharedCache->rememberClass((TR_OpaqueClassBlock *)ramClasses[0]);
```
or use `isPointerInSharedCache` which doesn't assert and fail the compiler if the output param is NULL.

AlexeyKhrabrov commented:
> or use `isPointerInSharedCache` which doesn't assert and fail the compilation if the output param is NULL.

This seems like the best course of action for this PR, at least for now. It will match what `J9::AheadOfTimeCompile::getClassChainOffset()` does in the same situation.

The only other place in the JIT that I could find where the result of `rememberClass()` is (almost) immediately passed to `offsetInSharedCacheFromPointer()` is in `TR::SymbolValidationManager::populateWellKnownClasses()`: https://github.com/eclipse/openj9/blob/master/runtime/compiler/runtime/SymbolValidationManager.cpp#L246-L264. It could also be changed to use `isPointerInSharedCache()` and fail the compilation if it fails.

hangshao0 commented:
> The only other option that would make the storeSharedData API stronger would be if it held the SCC write lock until commitUpdate finished. Is that doable

It is already the case. We have an assertion that the current thread has the write mutex after commitUpdate().

See: 
https://github.com/eclipse/openj9/blob/834cc45a3e7abc4ff571e5c4f4a2258b43110093/runtime/shared_common/CacheMap.cpp#L3822-L3828

dsouzai commented:
Ah, that's unfortunate. Thanks for the info Hang. @AlexeyKhrabrov I guess just go ahead with the plan to use `isPointerInSharedCache`.

AlexeyKhrabrov commented:
I've updated the code to avoid using `offsetInSharedCacheFromPointer()`. I've also opened a separate PR #12550 that fixes the same potential issue in `populateWellKnownClasses()`.

genie-openj9 commented:
Can one of the admins verify this patch?

mpirvu reviewed (commented):
Overall it looks good. I only have some small comments.
I would like to better understand synchronization issues: who calls `reset()` which sets/resets the `_resetInProgress` flag? Also, there are multiple monitors and some of them can be locked at the same time. Do we obey a strict order of acquiring those monitors to avoid deadlocks?

AlexeyKhrabrov commented on runtime/compiler/control/JITServerHelpers.cpp:
```diff
@@ -321,14 +320,18 @@ packROMClass(J9ROMClass *romClass, TR_Memory *trMemory)
 
-   ctx._packedRomClass = (J9ROMClass *)trMemory->allocateStackMemory(totalSize);
+   // Check if expected size matches, otherwise fail early
+   if (expectedSize && (expectedSize != packedSize))
```
Added a comment.

AlexeyKhrabrov commented on runtime/compiler/control/JITServerHelpers.cpp:
```diff
@@ -361,11 +364,11 @@ packROMClass(J9ROMClass *romClass, TR_Memory *trMemory)
    auto end = (uint8_t *)OMR::alignNoCheck((uintptr_t)ctx._cursor, sizeof(uint64_t));
-   TR_ASSERT(end == (uint8_t *)ctx._packedRomClass + totalSize, "Invalid final cursor position: %p != %p",
-             end, (uint8_t *)ctx._packedRomClass + totalSize);
+   TR_ASSERT(end == (uint8_t *)ctx._packedRomClass + packedSize, "Invalid final cursor position: %p != %p",
```
Done.

AlexeyKhrabrov commented on runtime/compiler/runtime/JITServerAOTDeserializer.cpp:
```diff
@@ -0,0 +1,854 @@
+      }
+   else
+      {
+      return deserializationFailure(method, comp, wasReset);
```
Fixed and refactored this method to have a single piece of code that adds the new IDs.

AlexeyKhrabrov commented on runtime/compiler/runtime/JITServerAOTDeserializer.cpp:
```diff
@@ -0,0 +1,854 @@
+      }
+
+   // Get the class name using the cached ROMClass SCC offset
+   J9ROMClass *romClass = _sharedCache->romClassFromOffsetInSharedCache(it->second._romClassSCCOffset);
```
`romClassFromOffsetInSharedCache()` already has a fatal assertion. Added a comment.

AlexeyKhrabrov commented on runtime/compiler/runtime/JITServerAOTDeserializer.cpp:
```diff
@@ -0,0 +1,854 @@
+
+   // Deserialize/validate and cache serialization records, keeping track of IDs of the new ones.
+   // Since the records are sorted in "dependency order", by the time a given record is about
+   // to be cached, all the records that is depends on are already successfully cached.
```
Fixed.

AlexeyKhrabrov commented:
> I would like to better understand synchronization issues: who calls `reset()` which sets/resets the `_resetInProgress` flag?

The `reset()` function will be called when the client detects that it connected to a new server instance. The best way to do this that I could think of is to use the `VM_getVMInfo` message. I haven't included it in this PR since it focuses on the infrastructure and not on the control logic yet. I updated the comment to make the purpose of this function more clear.

> Also, there are multiple monitors and some of them can be locked at the same time. Do we obey a strict order of acquiring those monitors to avoid deadlocks?

Yes. I've added a comment describing the locking hierarchy.

mpirvu reviewed (approved):
LGTM

mpirvu commented:
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit jdk11

mpirvu commented:
jenkins test sanity xlinuxjit jdk11

mpirvu commented:
`jit_jitt_array_compress_3` is the only test that fails (multi-platform) and this is a known issue.
